Add Prusa Printer Control Extension

chadrwalters opened the pull request:
## What does this PR do?
Adds a new extension to control Prusa 3D printers directly from Raycast. Supports monitoring printer status, managing print files, and controlling print jobs through PrusaLink API.

### Features
- 🖨️ Live printer status monitoring (temperatures, progress, state)
- 📁 File management with search and sorting
- 🎛️ Print control actions (pause/resume/cancel)
- 🔄 Automatic retry for network issues
- 🚀 Keyboard-driven interface

## Supported Printers
- Prusa XL (built-in PrusaLink)
- Prusa MK4 (built-in PrusaLink)
- Prusa MINI+ (built-in PrusaLink)
- Prusa MK3S+ (via Raspberry Pi running PrusaLink)

## Screenshots
![prusa-1](https://github.com/user-attachments/assets/24a93319-14a5-42a2-8f1b-262c1cf3ac8a)
![prusa-2](https://github.com/user-attachments/assets/5d01d0ba-e6aa-4093-96b3-6db6690ace8b)
![prusa-3](https://github.com/user-attachments/assets/f6f68aa0-c4dd-4754-95c8-ff030c4bf5ce)


## How to test
1. Install extension
2. Configure printer:
   - Enter printer IP address
   - Add PrusaLink API key from printer's web interface
3. Test commands:
   - "Printer Status" - View live printer information
   - "Print Files" - Browse and manage print files

## Additional notes
- Uses PrusaLink API v1
- Requires printer to be on the same network
- Includes automatic retry logic for network issues
- Supports file thumbnails when available from printer

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented on extensions/prusa/CHANGELOG.md:
```diff
@@ -0,0 +1,41 @@
+  - Configurable request timeout
+  - Input validation
+
+### Supported Printers
+- Prusa XL (built-in PrusaLink)
+- Prusa MK4 (built-in PrusaLink)
+- Prusa MINI+ (built-in PrusaLink)
+- Prusa MK3S+ (via Raspberry Pi running PrusaLink)
```
This doesn't belong here, made with AI?

pernielsentikaer reviewed (changes requested):
@greptileai can you check this

chadrwalters commented:
Thank you for the thorough review! I've addressed all the issues:

1. Removed the "Supported Printers" section from CHANGELOG.md as it doesn't belong there
2. Created a `createPrusaClientFromPreferences()` helper function in prusaClient.ts to centralize client initialization and validation logic
3. Updated both status.tsx and files.tsx to use the new helper function, eliminating code duplication
4. Fixed the retry logic in retry.ts:
   - Changed `attempt <= maxRetries` to `attempt < maxRetries` to fix the off-by-one error
   - Removed the redundant retry condition check
   - Improved error handling by removing unnecessary error initialization
5. Fixed the retry count logic in status.tsx to display the correct attempt number
6. Added status to the dependency array in useEffect to prevent stale state issues
7. Fixed all linting issues

All tests are passing, and the code is now more maintainable with better error handling. Let me know if you'd like any additional changes!

pernielsentikaer reviewed (commented):
@chadrwalters can you show me a screencast of how it works?

@greptileai can you check again

chadrwalters commented:
![NewProject](https://github.com/user-attachments/assets/113d32da-c0ee-40a8-a101-8e8a591c083c)

Here you go. I did not pause or start a print as I am in the middle of a print now.

pernielsentikaer reviewed (approved):
Hi 👋 

Looks good to me, approved 🔥
